domains/certificates/RegistryConnector/Worker/EventHandlers/ProductionCertificateCreatedEventHandler.cs

```diff
@@ -1,68 +1,92 @@
 using Microsoft.Extensions.Options;
-using NSec.Cryptography;
-using ProjectOrigin.Electricity.Client;
-using ProjectOrigin.Electricity.Client.Models;
-using RegistryConnector.Worker.Cache;
+using ProjectOrigin.HierarchicalDeterministicKeys.Implementations;
+using ProjectOrigin.PedersenCommitment;
+using ProjectOrigin.Registry.V1;
+using ProjectOriginClients;
 
 namespace RegistryConnector.Worker.EventHandlers;
 
 public class ProductionCertificateCreatedEventHandler : IConsumer<ProductionCertificateCreatedEvent>
 {
-    private readonly RegistryOptions registryOptions;
     private readonly ILogger<ProductionCertificateCreatedEventHandler> logger;
-    private readonly ICertificateEventsInMemoryCache cache;
-    private readonly IssuerKey issuerKey;
-    private readonly RegisterClient registerClient;
-
-    public ProductionCertificateCreatedEventHandler(IOptions<RegistryOptions> registryOptions,
-        RegisterClient registerClient,
-        ILogger<ProductionCertificateCreatedEventHandler> logger,
-        ICertificateEventsInMemoryCache cache,
-        IssuerKey issuerKey)
+    private readonly ProjectOriginOptions projectOriginOptions;
+
+    public ProductionCertificateCreatedEventHandler(
+        IOptions<ProjectOriginOptions> projectOriginOptions,
+        ILogger<ProductionCertificateCreatedEventHandler> logger)
     {
-        this.registryOptions = registryOptions.Value;
+        this.projectOriginOptions = projectOriginOptions.Value;
         this.logger = logger;
-        this.cache = cache;
-        this.issuerKey = issuerKey;
-        this.registerClient = registerClient;
     }
 
     public async Task Consume(ConsumeContext<ProductionCertificateCreatedEvent> context)
     {
-        var msg = context.Message;
-        var ownerKey = Key.Create(SignatureAlgorithm.Ed25519);
+        var message = context.Message;
+
+        var commitment = new SecretCommitmentInfo((uint)message.ShieldedQuantity.Value); //TODO: commitment should be part of message. See https://github.com/Energinet-DataHub/energy-origin-issues/issues/1572
+
+        var ownerKey = new Secp256k1Algorithm().GenerateNewPrivateKey(); //TODO: Derive new public key from Deposit Endpoint Reference. See https://github.com/Energinet-DataHub/energy-origin-issues/issues/1693
+        var ownerPublicKey = ownerKey.PublicKey;
+
+        var issuerKey = projectOriginOptions.GetIssuerKey(message.GridArea);
+
+        var issuedEvent = Registry.CreateIssuedEventForProduction(
+            projectOriginOptions.RegistryName,
+            message.CertificateId,
+            message.Period.ToDateInterval(),
+            message.GridArea,
+            message.Gsrn.Value,
+            commitment,
+            ownerPublicKey);
+
+        var request = issuedEvent.CreateSendTransactionRequest(issuerKey);
+
+        using var channel = GrpcChannel.ForAddress(projectOriginOptions.RegistryUrl); //TODO: Is this bad practice? Should the channel be re-used?
```

Comment:
Is this bad practice? Depends:

1. Making a new Grpc channel per transaction, create alot of overhead, can we handle it?
2. Security wise, it is more liable, if it is compromised. Should be safe for reuse, if proper encryption is in place.
3. If the time between issuances of certificates is long, maybe because we have introduced a waiting time, keeping the same channel open, could be a waste. If that is the case, then closing that channel, and creating a new one, could be more beneficial
4. Right now there doesn't seem to be any concurrency, as you have set a limit for that, but we might want to be able to do concurrent requests, then multiple channels should be created.
5. For error handling, if an exception is reached, maybe defaulting to creating a new channel, could be wise, so we are working with a clean slate?
6. As of now, transactions and issuances, seem more infrequent rather than frequent. Because of a slow datahub, and metering point owners, sending in updated metering point data, frequently. (Was it an up to 120 hour delay, we should expect?) It might make more sense to create new channel each time.

Reply:
Ok. Thanks for the input. Can this conversation be closed?